### DISCUSSION TITLE
Implementing undo (Ctrl+z) to my APIs

### DISCUSSION THREAD

**Discussion by ancelin-charpentier** [cwapi3d/cwapi3dpython#215] [Q&A] [answered]
Hey there,
I do have some APIs moving elements in my model. I want to keep it like that but if the command is started by accident then I got no options to restore the previous state.
Only solution I have now is saving before to start the API and to restart without saving...
Is there a way to allow undo (Ctrl+Z) to work for what an API did in the model?

Regards

**Comment by Antoine-S-C** [comment]
Hi,
you can use this element_controller function :
ec.add_created_elements_to_undo(elements_list)

**Reply by ancelin-charpentier** [reply]
I tried as well with
"element_controller.add_elements_to_undo(elements: List[int], **cmd: int**)"

but the 2nd entryhas to be an int for a command, I don't get that... Does command have an Int?

**Comment by ancelin-charpentier** [comment]
Hi Antoine,
Thanks for your response. It works very well for whatever is created by the API.
I would like the same but for elements that were just moved by the API. 
I tried with :
element_controller.add_modified_elements_to_undo(moved_part)

As a result the moved part disappears. They do not regain their previous position. Do you have a clue?

Cheers

**Reply by Antoine-S-C** [reply]
Hello,
for moved elements you can try :
element_controller.move_element_with_undo(elements, vector)
Let me know if it works.

**Reply by ancelin-charpentier** [reply]
Hello, thanks for that it worked well for the moved elements.
The API i wanted to get the undo added to is also using "element_controller.apply_transformation_coordinate". 
I wonder if there is any ways to undo this command. 
Maybe dumb suggestion there but could "element_controller.add_elements_to_undo(elements: List[int], cmd: int)" had a 3rd option for "move" @Brunner246 ?

Thanks for your help

**Reply by Brunner246** [reply, marked as answer]
You could do this 

```python
if __name__ == '__main__':
    active_element_ids = ec.get_active_identifiable_element_ids()
    if len(active_element_ids) == 0:
        logging.info("No active elements found")
        sys.exit(1)

    reference_element_id = active_element_ids[0]
    
    # you can do this 
    # ec.add_modified_elements_to_undo(active_element_ids)
    # or this - cmd argument => 2 = modify
    ec.add_elements_to_undo(active_element_ids, 2)

    ec.apply_transformation_coordinate(active_element_ids, gc.get_p1(reference_element_id),
                                       gc.get_xl(reference_element_id), 
                                       gc.get_yl(reference_element_id), 
                                       cadwork.point_3d(0., 0., 0.), 
                                       cadwork.point_3d(1., 0., 0.), 
                                       cadwork.point_3d(0., 1., 0.))
```

**Reply by ancelin-charpentier** [reply]
Hey @Brunner246,
I works perfectly thanks! 
Can't believe it was that easy...
Any chance to have your look onto the other query I have regarding right-click vs displayed menu from API?

**Comment by Brunner246** [comment]
Hi @ancelin-charpentier 

Sorry, we haven't documented this properly yet.
Possible Cmd values (int value) are: 

```
1 = Add
2 = Modify
```

**Reply by ancelin-charpentier** [reply]
Hi @Brunner246 , thanks for your reply, could you check my response to Antoine above? I guess you might be able to answer?

**Reply by Brunner246** [reply]
There is no 3rd option. There are just Add and Modify available.
If you want to move with undo, you have to use 
```python
element_controller.move_element_with_undo(List[elementIDs], cadwork.point_3d(1200., 0,  500.))
```

But you could create your own wrapper to treat the undo functions equally.
Maybe something like:
```python
class UndoType(Enum):
    ADD = 1
    MODIFY = 2
    MOVE = 3
    DELETE = 4

    
def add_elements_to_undo(element_ids: List[int], undo_type: UndoType):
    if undo_type == UndoType.ADD:
        ec.add_elements_to_undo(element_ids, UndoType.ADD.value)
    elif undo_type == UndoType.MODIFY:
        ec.add_elements_to_undo(element_ids, UndoType.MODIFY.value)
    elif undo_type == UndoType.MOVE:
        ec.move_element_with_undo(element_ids)
    elif undo_type == UndoType.DELETE:
        ec.delete_elements_with_undo(element_ids)
```
